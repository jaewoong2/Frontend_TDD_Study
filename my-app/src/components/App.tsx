import useChange from 'hooks/useChange';
import useLists from 'hooks/useLists';
import styled from 'styled-components';
import { Forms } from './blocks/Forms';
import Lists from './blocks/Lists';

function App() {
  const [todo, updateState, onChangeTodo] = useChange('');
  const { lists, addList, removeList } = useLists([]);

  return (
    <Container>
      <Contents>
        <Lists lists={lists} handleDelete={removeList} />
        <Forms
          removeTodo={() => updateState('')}
          handleAddTodo={addList(todo)}
          onChangeTodo={onChangeTodo}
          todo={todo}
          placeHolder="할 일을 입력해 주세요"
        />
      </Contents>
    </Container>
  );
}

export default App;

const Container = styled.div`
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const Contents = styled.div`
  display: flex;
  flex-direction: column;
  background-color: #fff;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
`;
