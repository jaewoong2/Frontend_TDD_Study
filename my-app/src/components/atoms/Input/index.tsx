import { forwardRef } from 'react';
import styled from 'styled-components';

export const InputContainer = styled.label`
  display: flex;
`;

const StyledInput = styled.input`
  font-size: 16px;
  margin-right: 9px;
  padding: 10px 10px;
  border-radius: 8px;
  border: 1px solid #bdbdbd;
  outline: none;
  flex: 1;
`;

type Props = React.InputHTMLAttributes<HTMLInputElement>;

export const Input = forwardRef<HTMLInputElement, Props>(
  ({ value, onChange, placeholder }, ref) => {
    return (
      <StyledInput
        value={value}
        onChange={onChange}
        placeholder={placeholder}
        ref={ref}
      />
    );
  },
);

Input.displayName = 'Input';
Input.defaultProps = {
  onChange: () => {},
};
