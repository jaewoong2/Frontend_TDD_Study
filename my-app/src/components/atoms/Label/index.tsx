import styled from 'styled-components';

export const Label = styled.label`
  color: #fff;
  font-size: 16px;
`;
