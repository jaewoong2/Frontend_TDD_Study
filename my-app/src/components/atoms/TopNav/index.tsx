import React from 'react';
import styled from 'styled-components';

type Props = {
  children?: React.ReactNode;
  left?: React.ReactNode;
  right?: React.ReactNode;
};

function TopNav({ children, left, right }: Props) {
  return (
    <Nav>
      <div className="box left">{left}</div>
      <div className="box center">{children}</div>
      <div className="box right">{right}</div>
    </Nav>
  );
}

export { TopNav };

const Nav = styled.nav`
  width: 100%;
  height: 40px;
  background-color: #9bceee;
  display: flex;
  border-radius: 8px;
  justify-content: center;
  align-items: center;

  .box {
    width: 100%;
  }

  .center {
    display: flex;
    justify-content: center;
    align-items: center;
  }
`;
