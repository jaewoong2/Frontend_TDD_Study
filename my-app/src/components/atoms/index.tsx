export { Button } from './Button';
export { Input, InputContainer } from './Input';
export { Item } from './Item';
