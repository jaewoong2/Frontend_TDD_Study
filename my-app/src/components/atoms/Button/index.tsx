import React from 'react';
import styled from 'styled-components';
import { Label } from '../Label';

type Props = {
  backgroundColor?: string;
  hoverColor?: string;
} & React.DetailsHTMLAttributes<HTMLButtonElement>;

export const Button = React.forwardRef<HTMLButtonElement, Props>(
  ({ backgroundColor, hoverColor, children, onClick }, ref) => {
    return (
      <Btn
        backgroundColor={backgroundColor}
        onClick={onClick}
        hoverColor={hoverColor}
        ref={ref}>
        <Label>{children}</Label>
      </Btn>
    );
  },
);

Button.displayName = 'Button';
Button.defaultProps = {
  backgroundColor: '#304FFE',
  hoverColor: '#1E40FF',
};

const Btn = styled.button<Props>`
  outline: none;
  border: none;
  text-align: center;
  background-color: ${({ backgroundColor }) => backgroundColor};
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
  &:hover {
    background-color: ${({ hoverColor }) => hoverColor};
  }
  &:active {
    box-shadow: inset 5px 5px 10px #00000033;
  }
`;
