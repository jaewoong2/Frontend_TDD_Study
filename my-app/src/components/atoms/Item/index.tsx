import styled from 'styled-components';
import React from 'react';
import { Label } from '../Label';
import { Button } from '../Button';

type Props = {
  children: React.ReactNode;
  onClick?: () => void;
  hoverColor?: string;
  backgroundColor?: string;
  buttonName?: string;
};

function Item({
  children,
  onClick,
  hoverColor,
  backgroundColor,
  buttonName,
}: Props) {
  return (
    <ItemContainer>
      <Label>{children}</Label>
      <Button
        onClick={onClick}
        hoverColor={hoverColor}
        backgroundColor={backgroundColor}>
        {buttonName}
      </Button>
    </ItemContainer>
  );
}

export { Item };

Item.defaultProps = {
  onClick: () => {
    console.log('onClick Button');
  },
  hoverColor: '',
  buttonName: '',
  backgroundColor: '',
};

const ItemContainer = styled.li`
  display: flex;
  align-items: center;
  border-bottom: 1px solid #bdbdbd;
  margin: 10px;
  padding: 10px;

  label {
    flex: 1;
    color: black;
  }
`;
