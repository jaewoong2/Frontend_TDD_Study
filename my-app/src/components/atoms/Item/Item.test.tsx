import { fireEvent, render, screen } from '@testing-library/react';
import { Item } from 'components/atoms';
import 'jest-styled-components';

describe('<Input />', () => {
  it('renders component correctly', () => {
    const { container } = render(
      <Item buttonName="Delete">Default Value</Item>,
    );

    const label = screen.getByText('Default Value');
    expect(label).toBeInTheDocument();
    const Button = screen.getByText('Delete');
    expect(Button).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('click delete button', () => {
    const handleClick = jest.fn();
    render(
      <Item buttonName="Delete" onClick={handleClick}>
        Default Value
      </Item>,
    );

    const deleteButton = screen.getByText('Delete');
    expect(handleClick).toHaveBeenCalledTimes(0);
    fireEvent.click(deleteButton);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
