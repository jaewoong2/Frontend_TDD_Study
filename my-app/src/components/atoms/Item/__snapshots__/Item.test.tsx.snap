// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> renders component correctly 1`] = `
.c1 {
  color: #fff;
  font-size: 16px;
}

.c2 {
  outline: none;
  border: none;
  text-align: center;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
}

.c2:active {
  box-shadow: inset 5px 5px 10px #00000033;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-bottom: 1px solid #bdbdbd;
  margin: 10px;
  padding: 10px;
}

.c0 label {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  color: black;
}

<div>
  <li
    class="c0"
  >
    <label
      class="c1"
    >
      Default Value
    </label>
    <button
      class="c2"
    >
      <label
        class="c1"
      >
        Delete
      </label>
    </button>
  </li>
</div>
`;
